### PR TITLE
[no-ci] [precompile] Mask guarded literals out of rendered guard code

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #197353
* #197352
* #197347
* #197345
* #197344
* #197456
* #197763
* #197767
* #197766
* #197343
* #197351
* #197839
* #197998
* __->__ #197849
* #197844
* #197689
* #197720
* #197600
* #197388
* #197387
* #197455
* #197341
* #197342
* #197348
* #197349

**Summary.** A guard that pins a string pins it BY VALUE, so the check `guards.py` renders carries the string itself -- an `EQUALS_MATCH` on a system prompt renders as `L['self'].prompt == 'you are ...'` and a dict keyed by a checkpoint path renders that path inside the guard's own name -- and the guard reports built on these renderings are meant to be committed to a file and diffed, which rules out printing a value. This commit makes `_render_code`, the module's renderer for a guard's check parts, mask the values a check embeds by type before `_normalize` scrubs the addresses out of it, and adds `_mask_keys` for the data key a source NAME interpolates (`cfg['/home/me/w.pt']` -> `cfg[<str>]`). Masking is an `ast.NodeTransformer`: a `str` or `bytes` constant becomes `<str>`/`<bytes>` and a whole `List`/`Set`/`Dict`/`Tuple` display becomes `<list>`/`<set>`/`<dict>`/`<tuple>`, while numbers, bools and `None` stay, since the number IS the check for a length or a shape and none of them is a value a report can leak. What makes a report auditable is the SHAPE of the check and the slot it names, so the deliberate keeps are the things that spell a source rather than a value: a name-like subscript key (`L['x']`, `G['CFG']`, `___dict__['act']`, and a normalized global's `_<id>_c<n>` placeholder) and the attribute name a `getattr`/`hasattr` reads. A check the pass finds nothing to mask in comes back exactly as `guards.py` wrote it rather than round-tripped through `ast.unparse`, the `, type=<class 'int'>` tail `___check_type_id` appends is split off around the parse, and a rendering that is not an expression at all -- the saved-tensors-hooks guard renders prose -- falls back to masking quoted runs textually.

## Why this slice

The rendering layer only: pure functions over one rendered check and one source name, with no session state and no accounting of what a capture kept or dropped. The commit above is what records those verdicts and is `_render_code`'s first library caller, so `_mask_keys` has no library caller yet and is exercised here by tests; it is applied to the recorded slot name one commit up. Masking has to run BEFORE `_normalize` -- the `<id>` and `<n>` placeholders it interpolates no longer parse as Python, and a name-like key it rewrites would read as a quoted value to the textual fallback -- which is why the rules live with `_render_code` rather than in the recorder, and why `GuardFact.code`'s contract is stated here.

Test Plan:

`TestGuardCodeMasking` drives the masking rules directly: each shape a value takes in a rendered check (a bare string and a bytes pin, a container display, a keys list, a containment argument, and a data key both in a subscript and in a slot name), each deliberate keep (a number, a bool, `None`, a name-like key, the attribute name a `hasattr` reads), the annotation tail coming off and going back on, the non-expression fallback, and the mask-before-normalize order against a rendering that carries both a builtins counter and an install_global_by_id name. Two of its tests run real captures of modules that pin a prompt, a path, a set, a keys list and a dict key, assert the rendering each guard type produces, and end by asserting that no pinned literal reaches any rendered check or any masked slot name.

```
python test/dynamo/test_precompile_package.py -k TestGuardCodeMasking
python test/dynamo/test_precompile_package.py
lintrunner -a torch/_dynamo/precompile_package.py torch/compiler/_precompile_types.py test/dynamo/test_precompile_package.py
```

Authored with an AI assistant.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude please review head 2a62495a921 (first review of this PR; it is the lower half of #197839, whose round-2 run hit the time ceiling at 838 inserted lines).

**Comment by bobrenjc93** [comment]
Thanks -- the layer now fails closed. Dispositions against head 96e48b2e34a.

> The SyntaxError fallback masks nothing when the repr carries no quotes (live leak).

Fixed in 96e48b2e34a: a check that does not parse renders as a fixed placeholder; no quoted-run substitution remains.

> `_SLOT_KEY` cannot distinguish a Dynamo scope key from a user dict key.

Fixed in 96e48b2e34a: a subscript key is kept only when its base is keyed by name (`L`, `G`, `___dict__`, the `nn.Module` name dicts `__dict__`/`_modules`/`_parameters`/`_buffers`, the builtins dict) and the key is name-shaped; every other key masks. `self.cfg['api_key']` and `L['cfg']['api_key']` mask, `L['x']`, `G['CFG']` and `_modules['lin']._parameters['weight']` keep (a base-only rule would have masked every parameter slot name).

> `getattr`'s default value and all keywords are never visited.

Fixed in 96e48b2e34a: only the attribute-name argument of `getattr`/`hasattr` is exempt; the object, the default and all keywords are visited.

> Whole-display masking collides genuinely different variants.

Fixed in 96e48b2e34a: displays still mask whole (the report must not carry user data) but the placeholder keeps the element count, so different-length containers stay distinct; the residual collision is stated in the docstring.

> `_mask_values` is not idempotent, and a second application destroys every slot name.

Fixed in 96e48b2e34a: the placeholder is a valid constant node; a second application is a no-op, tested.

> `except SyntaxError` doesn't cover `ast.unparse`.

Fixed in 96e48b2e34a: an f-string masks to one placeholder (the value is the joined string, and a format spec is itself an f-string), and the unparse path is guarded defensively; on 3.12 `ast.unparse` does not actually refuse the piecewise-masked tree, which the comment now says.

> `_CHECK_ANNOTATION` is `\Z`-anchored only.

Fixed in 96e48b2e34a: stripped wherever it appears, tolerant of quotes inside the class repr.

> `node.value = self.visit(node.value)  # type: ignore[assignment]`

Fixed in 96e48b2e34a: no ignores; the transformer returns fresh nodes consistently.

> `GuardFact.value` is not covered by the new contract.

Fixed in 96e48b2e34a (prose): the docstring states what `value` carries and why it is not masked; the claim was checked against main's `_value_fingerprint`.

> Gaps in an otherwise strong suite.

Fixed in 96e48b2e34a: the cases named under Testing are added and mutation-checked.

**Comment by bobrenjc93** [comment]
@claude please review the new head 96e48b2e34a.

**Comment by bobrenjc93** [comment]
Thanks -- one rule, one implementation now, and every unrecognized shape fails closed. Dispositions against head 18409823e81.

> `_QUOTED_KEY` only matches a slice that is exactly one quoted string, so a tuple key leaks its strings into the slot name.

Fixed in 18409823e81: the regex path is gone. Slot names are parsed and run through the same AST pass as the checks, so tuple, nested-tuple and frozenset keys mask (`cfg['<tuple:2>']`), numeric indices stay, and a name that is not an expression becomes `<unparsed source>`.

> The text and AST spellings of "keyed by a name" are not the same rule, and the text one is the looser.

Fixed in 18409823e81: the text spelling and its regexes are deleted; a shared `_mask_expr` serves both. While unifying, the AST rule's own `__builtins_dict__`-prefix hole was found and closed (it now recurses into the base).

> The comment justifying `_UNPARSED_CHECK` names a mitigation that does not exist for the guard type most likely to hit it.

Fixed in 18409823e81 (prose): the comment states what is true (an `EQUALS_MATCH` on a repr has no discriminator once dropped whole) and why no digest is written: a stable fingerprint of a hidden value in a committed file is an oracle for guessing it.

> Masking a display whole still costs discrimination it does not need to.

Declined (documented): displays nest without limit, so a rule that inspects elements to stay safe is defeated by an element shape it does not model; the docstring states the cost.

> `_SAVED_HOOK_IDS` is now unreachable from `_render_code`.

Fixed in 18409823e81: deleted, with its `_normalize` line and unit case; routing the hooks text through it first cannot help because the text is still not an expression afterwards, so it fails closed as a whole. A comment says so.

> The annotation is reattached at the end rather than in place.

Fixed in 18409823e81: the split position is recorded and the annotation returns where it sat; two annotations or a mid-body one fail closed. Stripping it entirely was rejected because the per-process type-id masking depends on the `, type=` lookahead.

> The three new findings are each one test away.

Fixed in 18409823e81: the three cases added, each mutation-checked.

**Comment by bobrenjc93** [comment]
@claude please review the new head 18409823e81.

**Comment by bobrenjc93** [comment]
Thanks -- every string now masks, whatever its shape. Dispositions against head efeceda104b.

> `_MASKED_VALUE` recognizes a placeholder by SHAPE, and ordinary strings have that shape -- so they are returned verbatim.

Fixed in efeceda104b: the shape sentinel is deleted; every string constant masks (`<pad>` and a user `'<list:3>'` both become `'<str>'`, as value and as key). Re-masking rewrites a typed placeholder downward to `'<str>'`, never upward, which the docstring states.

> The exemption covers `getattr`/`hasattr` only, but that is not the only rendering that puts an attribute NAME in a call.

Fixed in efeceda104b: a small table of callable -> name-argument index drives the exemption, with the dict-contains entry gated on a name-keyed base so a user dict's key still masks.

> `_mask_expr` narrows to `SyntaxError` and `ValueError`, leaving the traversal unguarded.

Fixed in efeceda104b: parse, traversal and unparse fail closed on any exception (a recursion error was reproduced before and after).

> Two pieces of the keep-rule are dead or stale on the path that now exists.

Fixed in efeceda104b: `___dict__` removed from the scope set; the stale comment now names the real caller.

> `GuardFact.source` and `PrecompileSummary`'s slot paragraph still describe an unmasked spelling.

Fixed in efeceda104b: both state the masked spelling, the kept keys and the placeholder.

> Three of the above are one case each.

Fixed in efeceda104b: the three tests added, each mutation-checked.

**Comment by bobrenjc93** [comment]
@claude please review the new head efeceda104b.

**Comment by bobrenjc93** [comment]
Closing: the author is taking a different direction with this work. Nothing here landed; the branch is left in place.
